test/replication-luatest/gh_6036_qsync_order_test.lua

```diff
@@ -0,0 +1,193 @@
+    -- sure the PROMOTE has reached it via queue state test.
+    cg.r2:exec(function()
+        box.ctl.promote()
+        box.ctl.wait_rw()
+    end)
+    t.helpers.retrying({}, function()
+        t.assert(cg.r3:exec(function()
+            return box.info.synchro.queue.busy == true
+        end))
+    end)
+    cg.r2:exec(function()
+        box.space.test:insert{2}
+    end)
+
+    --
+    -- The r1 node has no clue that there is a new leader and continue
+    -- writing data with obsolete term. Since r3 is delayed now
+    -- the INSERT won't proceed yet but get queued.
+    cg.r1:exec(function()
+        box.space.test:insert{3}
+    end)
+
+    --
+    -- Finally enable r3 back. Make sure the data from new r2 leader get
+    -- writing while old leader's data ignored.
+    cg.r3:exec(function()
+        box.error.injection.set('ERRINJ_WAL_DELAY', false)
+    end)
+    t.helpers.retrying({}, function()
+        t.assert(cg.r3:exec(function()
+            return box.space.test:get{2} ~= nil
+        end))
+    end)
+
+    t.assert_equals(cg.r3:exec(function() return box.space.test:select() end), {{1},{2}})
+end
+
+--
+-- Drop the r3 replica, since it is no longer needed for this test.
+g.after_test("test_qsync_order", function(cg)
+    cg.box_cfg.replication[3] = nil
+    cg.r1:exec(update_replication, cg.box_cfg.replication)
+    cg.r2:exec(update_replication, cg.box_cfg.replication)
+    cg.r3:stop()
+    cg.r3:cleanup()
+    cg.r3 = nil
+end)
+
+g.test_promote_order = function(cg)
+    --
+    -- Make sure that while we're processing PROMOTE no other records
+    -- get sneaked in via applier code from other replicas. For this
+    -- sake initiate voting and stop inside wal thread just before
+    -- PROMOTE get written. Another replica sends us new record and
+    -- it should be dropped.
+    cg.r1:exec(function()
+        box.ctl.promote()
+        box.ctl.wait_rw()
+    end)
+    local vclock = cg.r1:get_vclock()
+    vclock[0] = nil
+    cg.r2:wait_vclock(vclock)
+
+    --
+    -- Drop connection between r1 and the rest of the cluster.
+    -- Otherwise r1 might become Raft follower before attempting
+    -- insert{4}.
+    cg.r1:exec(function() box.cfg{replication=""} end)
+    cg.r2:exec(function()
+        box.error.injection.set('ERRINJ_WAL_DELAY_COUNTDOWN', 2)
+        require('fiber').create(function() box.ctl.promote() end)
+    end)
+    t.helpers.retrying({}, function()
+        t.assert(cg.r2:exec(function()
+            return box.info.synchro.queue.busy
+        end))
+    end)
+    t.assert(cg.r1:exec(function() return box.info.ro == false end))
+    cg.r1:exec(function()
+        box.space.test:insert{4}
+    end)
+    cg.r2:exec(function()
+        require('luatest').assert(box.info.synchro.queue.busy == true)
+        box.error.injection.set('ERRINJ_WAL_DELAY', false)
+        box.ctl.wait_rw()
+    end)
+
+    t.assert_equals(cg.r2:exec(function() return box.space.test:select() end), {{1},{2}})
```

Comment:
You assume that this test case runs after `test_qsync_order`, but I don't remember luatest giving any guarantees on the order of tests. Better not make the cases depend on each other's order.

Reply:
Good point. I'm not sure whom to ask. @sergepetrenko ?

Reply:
You shouldn't depend on the test order. Let's truncate the space between the test runs.
And assert that it is empty here.